lib/data/json.js

```diff
@@ -0,0 +1,201 @@
+
+            // if we have exactly reached our target table branch, push our new iteration
+            // to the final result and attach a parent id reference.
+            if (branchState === 0) {
+              result.push(bag);
+
+              // leverage the zipped contextStack we already have; drop one entry, then
+              // continue dropping until we get the next repeat.
+              do contextStack.pop();
+                while ((contextStack.length > 0) && (ptr(contextStack)[0].type !== 'repeat'));
+
+              // now push the relevant id.
+              if (contextStack.length === 0)
+                bag['__Submissions-id'] = submission.instanceId;
+              else
+                bag[`__${contextStack.map((ctx) => ctx[0].name).join('-')}-id`] = hashId(contextStack);
+            }
+          } else {
+            // now reset context so subtables are not emitted.
+            fieldStack.pop();
+            fieldStack.push(null);
+            dataStack.push(null);
+            iterationStack.push(null);
+          }
+        } else {
+          // for primitive fields, we iterate in-place; the value should be written into
+          // the current pointer position.
+          pushPtr(dataStack);
+          pushPtr(iterationStack);
+        }
+      } else {
+        // if we don't have a schema definition for this field, simply navigate into
+        // nothing; we still push stack state to track tree depth.
+        fieldStack.push(null);
+        dataStack.push(null);
+        iterationStack.push(null);
+      }
+    },
+    ontext: (text) => {
+      const fieldPtr = ptr(fieldStack);
+      if ((fieldPtr != null) && (fieldPtr.name != null)) {
+        if (getBranchState(fieldStack, table) === 1) {
+          const dataPtr = ptr(dataStack);
+          // we have a value and a place to put it. preprocess it if necessary and write.
+          if ((fieldPtr.type === 'structure') || (fieldPtr.type === 'repeat')) {
+            return;
+          } else if (fieldPtr.type === 'int') {
+            dataPtr[fieldPtr.name] = parseInt(text);
+          } else if (fieldPtr.type === 'decimal') {
+            dataPtr[fieldPtr.name] = parseFloat(text);
+          } else if (fieldPtr.type === 'geopoint') {
+            // all formats require this parsing/formulation:
+            const [ lat, lon, altitude ] = text.split(/\s+/g).map(parseFloat);
+            if ((lat == null) || (lon == null)) return;
+            if (Number.isNaN(lat) || Number.isNaN(lon)) return;
+            const coordinates = [ lon, lat ];
+            if ((altitude != null) && !Number.isNaN(altitude)) coordinates.push(altitude);
+
+            if (options.wkt === true) // well-known text format:
+              dataPtr[fieldPtr.name] = `POINT (${coordinates.join(' ')})`;
+            else // geojson is the default:
+              dataPtr[fieldPtr.name] = { type: 'Point', coordinates };
+          } else {
+            dataPtr[fieldPtr.name] = text;
```

Comment:
media??

Reply:
it'll just be opaque file references for now. eventually we can join against the attachments table and provide cogent URLs at least, even if the auth around said URLs is ??